File: Arrays/Easy/moveZeros.py
def moveZero(arr):
    n = len(arr)
    temp = [0]*n
    
    j = 0
    
    for i in range(n):
        if arr[i]!=0:
            temp[j] = arr[i]
            j+=1
            
    while j<n:
        temp[j] = 0
        j+=1
        
    for i in range(n):
        arr[i] = temp[i]
    
    return arr
     
    


# def moveZerosOptimized(arr:List[int],  n: int) -> int:
    j = -1
    # Place the pointer j
    for i in range(n):
        if arr[i] == 0:
            j = i
            break
    
    # No non-zero elements
    if j == -1:
        return arr
    
    # Move the pointers i and j and swap accordingly
    for i in range(j + 1, n):
        if arr[i] != 0:
            arr[i], arr[j] = arr[j], arr[i]
            j += 1
    
    return arr


arr = [1, 0, 2, 3, 2, 0, 0, 4, 5, 1]
moveZero(arr)
print(arr)
# n = 10
# ans = moveZerosOptimized(arr, n)
# for it in ans:
#     print(it, end=' ')
# print()




# arr = [1,0,2,0,3,0,4,5]
# n = len(arr)
# print(moveZero(arr,n))

File: Arrays/Easy/moveZeros.cpp
// Input: 1 ,0 ,2 ,3 ,0 ,4 ,0 ,1
// Output: 1 ,2 ,3 ,4 ,1 ,0 ,0 ,0

#include<iostream>
#include<vector>
using namespace std;


void moveZeros(vector<int>& arr){
    vector<int>temp;
    
    for(int i=0;i<arr.size();i++){
        if(arr[i]!=0){
         temp.push_back(arr[i]);
        }
    }

    for(int i=0;i<temp.size();i++){
        arr[i] = temp[i];
    }

    for(int i = temp.size();i<arr.size();i++){
        arr[i] = 0;
    }


}

void moveZerosOptimized()



int main(){
   vector<int> arr = {1,0,2,0,3,0,4,0,5,0,6,0,7};
   moveZeros(arr);
   for(auto it : arr){
    cout<<it<<endl;
   }
}




//   int n = sizeof(arr) / sizeof(arr[0]);
      
    // cout<<moveZeros(arr);
//     for(int i=0;i<n;i++){
//     cout<<arr[i]<<endl;
//  }